R/agg_scale.R

```diff
@@ -0,0 +1,885 @@
+#'   year = 2011, value = 62
+#' )
+#' input_dt <- rbind(input_dt, input_dt_agg)
+#' \dontrun{
+#' output_dt <- scale(dt = input_dt,
+#'                    id_cols = c("location", "year"),
+#'                    value_cols = "value",
+#'                    col_stem = "location",
+#'                    col_type = "categorical",
+#'                    mapping = iran_mapping,
+#'                    collapse_missing = TRUE)
+#' }
+#' issues_dt <- identify_scale_dt_issues(dt = input_dt,
+#'                                       id_cols = c("location", "year"),
+#'                                       value_cols = "value",
+#'                                       col_stem = "location",
+#'                                       col_type = "categorical",
+#'                                       mapping = iran_mapping,
+#'                                       collapse_missing = TRUE)
+#'
+#' @export
+#' @rdname identify_agg_scale_dt_issues
+identify_agg_dt_issues <- function(dt,
+                                   id_cols,
+                                   value_cols,
+                                   col_stem,
+                                   col_type,
+                                   mapping,
+                                   agg_function = sum,
+                                   drop_present_aggs = FALSE) {
+
+  # Validate arguments ------------------------------------------------------
+
+  assert_agg_scale_args(dt, id_cols, value_cols, col_stem,
+                        col_type, mapping, agg_function, "agg")
+
+  # check `drop_present_aggs` argument
+  assertthat::assert_that(assertthat::is.flag(drop_present_aggs),
+                          msg = "`drop_present_aggs` must be a logical")
+
+  cols <- col_stem
+  if (col_type == "interval") {
+    cols <- paste0(col_stem, "_", c("start", "end"))
+  }
+  by_id_cols <- id_cols[!id_cols %in% cols]
+
+  groups <- identify_unique_groupings(dt, col_stem, col_type,
+                                      by_id_cols)
+
+  # Check for aggregates already present or overlapping intervals -----------
+
+  aggs_present_dt <- lapply(1:nrow(groups$unique_groupings), function(group_num) {
+
+    grouping <- subset_unique_grouping(dt, groups, group_num, col_stem,
+                                       col_type, by_id_cols)
+
+    check_dt <- NULL
+    if (col_type == "categorical") {
+      agg_tree <- create_agg_tree(mapping, grouping$unique_cols[[cols]],
+                                  col_type)
+      present_nodes <- identify_present_agg(agg_tree)
+
+      if (!is.null(present_nodes)) {
+        # expand the groupings dataset to show the present aggregate nodes
+        check_dt <- grouping$dt[, list(present = present_nodes),
+                                by = by_id_cols]
+        data.table::setnames(check_dt, "present", cols[1])
+      }
+    } else if (col_type == "interval") {
+      # TODO add check for when start of interval is exactly equivalent to end
```

Comment:
Do you want to do this now or later?

Reply:
I added it to #8 since I found this issue when the example Iran census data had an unknown age group with 'age_start' and 'age_end' equal to 999 but the functions didn't catch it.